[ElginTZM] iP

ElginTZM opened the pull request:
### A-CAT (Automated Chatbot Assistant for Tasks)

> “Cats rule the world.” – Jim Davis ([source](https://www.shutterfly.com/ideas/cat-quotes/))

A-CAT is your ~~cat overlord~~ _personal assistant_ to manage your tasks! It's,

- very simple to use
- quick to setup
- text-based

To get started,

1.  Download A-CAT from [here](https://github.com/ElginTZM/ip/releases/tag/A-Jar)
2.  Double-click it
3.  Manage your tasks!

And it is **FREE**! 😺 

Features:
- [x] Manage Tasks, Deadlines and Events
- [x] Search for specific tasks
- [ ] Easy to use GUI

If you understand Java, you can take a look at the `main` method:
```java
public static void main(String[] args) {
    try {
        Stream<String> taskDataStream = Storage.readFile("tasks.txt");
        Duke.tasks = new TaskList(taskDataStream);

        if (tasks.hasLoadingError()) {
            Ui.showLoadingError();
        }

    } catch (DukeException e) {
        Ui.respond(e);
    }

    Ui.showGreetMessage();

    while (!isExit) {
        try {
            Command command = Ui.readCommand();

            if (command == null) {
                Ui.showInputArrow();
                continue;
            }

            command.execute(tasks, ui, storage);
            Duke.isExit = command.isExit();
        } catch (DukeException e) {
            Ui.respond(e);
        }
    }

    Ui.showExitMessage();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ElginTZM committed:
Update chatbot name, greeting and exit message

ElginTZM committed:
Add echo functionality to chatbot

ElginTZM committed:
Add adding and listing task functionality

ElginTZM committed:
Change String array to ArrayList<String>

ElginTZM committed:
Make respond functions generic

Respond function now takes in generic List<T> instead of List<String> and T instead of String.

ElginTZM committed:
Add Task class

ElginTZM committed:
Use Task class instead of storing as String

ElginTZM committed:
Add mark and unmark command to chatbot

ElginTZM committed:
Task class: make getStatusIcon protected

ElginTZM committed:
Add Task subclasses - Deadline, Todo, Event

ElginTZM committed:
Add todo, deadline and event command to chatbot

ElginTZM committed:
Fix bug allowing empty description for Todo

ElginTZM committed:
Update EXPECTED.TXT and input.txt

ElginTZM committed:
Add DukeException class

ElginTZM committed:
Update chatbot response style

ElginTZM committed:
Add exception handling for invalid inputs

ElginTZM committed:
Add delete command to chatbot

ElginTZM committed:
Add loading of task file on initial start

ElginTZM committed:
Fix task not being able to be marked undone

ElginTZM committed:
Task Class: Add getDataString method

ElginTZM committed:
Add addTask method

Remove addTodo, addDeadline, addEvent and respondWithAddedTask.
Add TaskType enum.

ElginTZM committed:
Update loadTasks and add storeTasks method

ElginTZM committed:
Merge branch 'branch-Level-7'

Commit:
Update how DateTime is stored in the Task classes

Commit:
Duke class: Add DateTime support for adding task

ElginTZM committed:
Merge branch 'branch-Level-8'

Commit:
Add Storage class

ElginTZM committed:
Update build.gradle mainClass

ElginTZM committed:
Merge branch 'add-gradle-support'

Commit:
Add Ui class

Commit:
Add TaskList class

Commit:
Add Parser class

Commit:
Add Command classes

Commit:
Task class: Update return types to enable chaining

Commit:
Update Duke and ExitCommand class

Commit:
Update test cases

ElginTZM committed:
Merge branch 'branch-A-MoreOOP'

ElginTZM committed:
Add duke.command package

ElginTZM committed:
Add duke.task package

ElginTZM committed:
Add duke.core package

ElginTZM committed:
Merge branch 'A-Packages'

ElginTZM committed:
Add TaskListTest

ElginTZM committed:
Update build.gradle

ElginTZM committed:
TaskListTest class: Remove unnecessary import

ElginTZM committed:
Add ParserTest

ElginTZM committed:
Merge branch 'A-JUnit'

ElginTZM committed:
Add FindCommand

ElginTZM committed:
Fix line break indentation

ElginTZM committed:
Add JavaDoc

ElginTZM committed:
Ui class: Add readCommand

ElginTZM committed:
Ui class: Add showGreetMessage and showExitMessage

ElginTZM committed:
Command class: Change constructor and add isExit

ElginTZM committed:
Command class: Add getParameterMap

ElginTZM committed:
Update parseCommand and parseTaskDataString

ElginTZM committed:
Duke class: update main running code

ElginTZM committed:
Improve command package coding standard

ElginTZM committed:
Ui class: add showLoadingError

ElginTZM committed:
Add support for showing loading tasks error

ElginTZM committed:
Ui class: Add showInputArrow and showSeparator

ElginTZM committed:
Storage class: improve imports formatting

ElginTZM committed:
Parser class: Fix indentation

ElginTZM committed:
Deadline class: Fix indentation

ElginTZM committed:
Event class: Fix indentation

ElginTZM committed:
ParserTest class: Reorder imports

ElginTZM committed:
TaskListTest class: Fix pokeball exception handling

ElginTZM committed:
Merge branch 'branch-A-CodingStandard'

ElginTZM committed:
Merge branch 'branch-A-JavaDoc'

ElginTZM committed:
Merge branch 'branch-Level-9'

ElginTZM committed:
Update build.gradle with checkstyle plugin

ElginTZM committed:
Add checkstyle configs

ElginTZM committed:
Fix styling issues detected by checkstyle

ElginTZM committed:
Merge branch 'A-CheckStyle'

ElginTZM committed:
Update build.gradle for javafx

ElginTZM committed:
Add Launcher class

ElginTZM committed:
Update build.gradle mainClass

ElginTZM committed:
Update return type of execute for Command

ElginTZM committed:
Add images for GUI

ElginTZM committed:
Storage class: Make methods non-static

ElginTZM committed:
Fix commands style

ElginTZM committed:
Add GUI and remove Ui class

tiongMax commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,147 @@
+    private TaskType taskType;
+
+    /** Enum to represent the type of task to be added. */
+    public enum TaskType {
```
Good job on using enum

tiongMax reviewed (commented):
Well done! Seems like u adhere to the coding standard well. Good job for applying knowledge from CS2030S! Keep it up!

AlagappanRa commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        if (tasks.size() == 0) {
+            throw new DukeException("There are no tasks added. Please add a task first.");
```
I like how all the exceptions are anticipated and well-handled! 😄

AlagappanRa reviewed (commented):
LGTM!

ElginTZM committed:
Merge branch 'branch-Level-10'

ElginTZM committed:
Add assertion to command classes

Command classes are not using java assertions to ensure certain assumptions are correct.

Add assertions to various points in AddCommand, DeleteCommand, MarkCommand and UnmarkCommand.

It is assumed that the TaskList class would return a valid task.

ElginTZM committed:
Improve abstraction of command classes and Parser

Command::execute handles reading, parsing and validating parameters.

This results in a very long execute function and has multiple layers of abstraction. This affects the readability of the code.

Add abstraction for loading, parsing and validating parameters that is called in the constructor.

When creating a new instance of the command, loading, parsing and validating can already be done instead of waiting until executing to check.

ElginTZM committed:
Merge pull request #4 from ElginTZM/branch-A-CodeQuality

Improve abstraction of command classes and Parser

ElginTZM committed:
Merge branch 'master' into branch-A-Assertions

ElginTZM committed:
Merge pull request #2 from ElginTZM/branch-A-Assertions

Add assertion to command classes

ElginTZM committed:
Add gradle.yml workflow file

ElginTZM committed:
Merge branch 'branch-A-CI'

ElginTZM committed:
Merge branch 'master' of https://github.com/ElginTZM/ip

ElginTZM committed:
Add sort command

ElginTZM committed:
Merge branch 'branch-C-Sort'

ElginTZM committed:
Make DialogBox text size smaller

ElginTZM committed:
Prevent resizing of A-CAT window

ElginTZM committed:
Update build.gradle to fix cross-platform issues

ElginTZM committed:
Fix not saving tasks when adding them

ElginTZM committed:
Fix checking of arguments for some commands

1.  FindCommand class: Make searching for keywords use lowercase instead
2. MarkCommand class: Throw an exception if no task number was supplied
3. UnmarkCommand class: Throw an exception if no task number was
supplied

ElginTZM committed:
Deadline class: Implement Comparable<Deadline>

ElginTZM committed:
SortCommand class: Add sorting by date

ElginTZM committed:
Merge branch 'branch-Enhancement'

ElginTZM committed:
Add user guide

ElginTZM committed:
Merge branch 'branch-A-UserGuide'
